python/cuml/common/array.py

```diff
@@ -0,0 +1,269 @@
+    case it'll keep a reference to that data underneath. Also can be created
+    from a pointer, specifying the characteristics of the array, in that case
+    the owner of the data referred to by the pointer should be specified
+    explicitly.
+
+    To standardize our code, please import this using:
+
+    from cuml.common.array import Array as cumlArray
+
+    Parameters
+    ----------
+
+    data : rmm.DeviceBuffer, array_like, int
+        An array-like object or integer representing a
+        device or host pointer to pre-allocated memory.
+    owner : object, optional
+        Python object to which the lifetime of the memory
+        allocation is tied. If provided, a reference to this
+        object is kept in this Buffer.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+    shape : int or tuple of ints, optional
+        Shape of created array.
+    order: string, optional
+        Whether to create a F-major or C-major array.
+
+    Attributes
+    ----------
+
+    ptr : int
+        Pointer to the data
+    size : int
+        Size of the array data in bytes
+    _owner : Python Object
+        Object that owns the data of the array
+    shape : tuple of ints
+        Shape of the array
+    order : {'F', 'C'}
+        'F' or 'C' to indicate Fortran-major or C-major order of the array
+    strides : tuple of ints
+        Strides of the data
+    __cuda_array_interface__ : dictionary
+        __cuda_array_interface__ to interop with other libraries.
+
+    Object Methods
+    --------------
+
+    to_output : Convert the array to the appropriate output format.
+
+    Class Methods
+    -------------
+
+    Array.empty : Create an empty array, allocating a DeviceBuffer.
+    Array.zeros : Create an Array with allocated DeviceBuffer initialized with
+        zeros.
+
+    Notes
+    -----
+
+    cuml Array is not meant as an end-user array library. It is meant for
+    cuML/RAPIDS developer consumption. Therefore it contains the minimum
+    functionality. Its functionality is hidden by base.pyx to provide
+    automatic output format conversion so that the users see the important
+    attributes in whatever format they prefer.
+
+    Todo: support cuda streams in the constructor, particularly after
+
+
+
+    """
+
+    def __init__(self, data=None, owner=None, dtype=None, shape=None,
+                 order=None):
+        if data is None:
+            raise TypeError("To create an empty Array, use the class method" +
+                            " Array.empty()")
+        if isinstance(data, int):
+            if dtype is None or shape is None or order is None:
+                raise TypeError("Need to specify dtype, shape and order when" +
+                                " creating an Array from a pointer.")
+
+        ary_interface = False
+
+        if isinstance(data, DeviceBuffer) or isinstance(data, int):
```

Comment:
Should we do anything special with `Buffer` instances that are not `Array`s? Asking as we may need to override, `shape`, `dtype`, `order`, and `strides` there as well.

Reply:
thanks to you actually, Buffer has cuda array interface support, so it'll be dealt correctly. If it is needed to something else with the Buffer, it should be pretty easy to view it differently by passing it to cupy, using`view` and then constructing an Array from it. If that becomes a need, we could even add that internally, not sure if it'll be needed for the time being though

Reply:
In that case, should we do the same thing with `DeviceBuffer` as it also supports [`__cuda_array_interface__`]( https://github.com/rapidsai/rmm/blob/0d62bf7aa7c2d105ef53b7a58b73a79992efe1bc/python/rmm/_lib/device_buffer.pyx#L36-L45 )?

Reply:
Actually after some thought yesterday, the inconsistency is better dealt with with the original suggestion I think. In general Buffer/DeviceBuffer types are lower level objects, that if we want to use to build arrays from (like I do in zeros, ones and full class methods), we usually will want to override their shape and rest of attributes

Reply:
Makes sense. Thanks Dante! 😄 Sorry if I wasn't grasping what was going on in this code before.